library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -242,6 +258,43 @@ evaluateTerm' ::
     EquationT io Term
 evaluateTerm' direction = iterateEquations 100 direction PreferFunctions
 
+{- | Simplify a Pattern, processing its constraints independently.
+     Returns either the first failure or the new pattern if no failure was encountered
+-}
+evaluatePattern ::
+    MonadLoggerIO io =>
+    Bool ->
+    KoreDefinition ->
+    Maybe LLVM.API ->
+    Pattern ->
+    io (Either EquationFailure (Pattern, [EquationTrace]))
+evaluatePattern doTracing def mLlvmLibrary =
+    runEquationT doTracing def mLlvmLibrary . evaluatePattern'
+
+-- version for internal nested evaluation
+evaluatePattern' ::
+    MonadLoggerIO io =>
+    Pattern ->
+    EquationT io Pattern
+evaluatePattern' Pattern{term, constraints} = do
+    pushConstraints constraints
+    newTerm <- evaluateTerm' TopDown term
+    -- after evaluating the term, evaluate all (existing and
+    -- newly-acquired) constraints, once
+    currentPredicates <- Set.toList . predicates <$> getState
+    traverse_ simplifyAssumedPredicate currentPredicates
```

Comment:
Can we use `mapM_` instead of `traverse_` here to avoid converting to the set to list? I.e.

```suggestion
    currentPredicates <- predicates <$> getState
    mapM_ simplifyAssumedPredicate currentPredicates
```

Reply:
Or actually, we could just use `traverse_` and not convert to list. Anyway, this is just a nitpick, feel free to ignore.

Reply:
Good point, using `traverse_` directly then.